lib/iris/tests/unit/fileformats/netcdf/test_save.py

```diff
@@ -222,37 +224,140 @@ class Test_HdfSaveBug(tests.IrisTest):
     is a specific problem, relating to HDF so limited to netcdf-4 formats.
     See : https://github.com/Unidata/netcdf-c/issues/1772
 
+    In all these testcases, a straightforward translation to the file would be
+    able to save [cube_2, cube_1], but *not* [cube_1, cube_2],
+    because the latter creates a dim of the same name as the 'cube_1' data
+    variable.
+
+    Here, we are testing the specific workarounds in Iris netcdf save which
+    avoids that problem.
+    Unfortunately, owing to the complexity of the iris.fileformats.netcdf.Saver
+    code, there are several separate places where this had to be fixed.
+
+    N.B. we also check that the data (mostly) survives a save-load roundtrip.
+    To identify the read-back cubes with the originals, we use var-names,
+    which works because the save code opts to adjust dimension names _instead_.
+
     """
 
-    def test_problem_case(self):
-        n_x = 2
-        x_dim = DimCoord(
-            np.arange(n_x), long_name="dim_x", var_name="same_name"
-        )
-        cube_x = Cube(np.arange(n_x), long_name="cube_x", var_name="same_name")
-        cube_y = Cube(
-            np.arange(n_x), long_name="cube_y", var_name="different_name"
-        )
-        cube_y.add_dim_coord(x_dim, 0)
-        # In this case, a straightforward translation to the file will be able
-        # to save [cube_y, cube_x], but *not* [cube_x, cube_y], because the
-        # latter makes a dim of the same name as the 'cube_x' data variable.
-        # Here, we are testing the specific workaround in Iris netcdf save which
-        # avoids that problem.
+    def _check_save_and_reload(self, cubes):
         tempdir = Path(mkdtemp())
         filepath = tempdir / "tmp.nc"
         try:
-            cubes = [cube_x, cube_y]
-            save(cubes, filepath)  # , netcdf_format='NETCDF3_CLASSIC')
-            # from subprocess import check_call
-            # check_call(f'ncdump -h {filepath}', shell=True)
-            loadback_cube_y = iris.load_cube(str(filepath), "cube_y")
-            self.assertEqual(
-                loadback_cube_y.coord("dim_x").var_name, "same_name_0"
-            )
+            # Save the given cubes.
+            save(cubes, filepath)
+
+            # Load them back for roundtrip testing.
+            with PARSE_UGRID_ON_LOAD.context():
+                new_cubes = iris.load(str(filepath))
+
+            # There should definitely still be the same number of cubes.
+            self.assertEqual(len(new_cubes), len(cubes))
+
+            # Get results in the input order, matching by var_names.
+            result = [new_cubes.extract_cube(cube.var_name) for cube in cubes]
```

Comment:
Could you add a clause to confirm that the cube ordering has indeed been altered - evidencing that a name clash has been handled?

The test cases in this class are all based on assumptions of how Iris will name variables. If these conventions are ever altered without this test class being altered too (seems likely), the tests will continue to pass but will not be providing the coverage they were written for - there won't be any name clashes that need handling!

I assume this suggestion would break the string test until #4412 is solved? Not sure what to do about that.

Reply:
@pp-mo has explained to me that a change in `Cube` ordering is not symptomatic of a name clash having been handled.

There is _always_ a chance that `Cube`s will load in a different order to which they were saved. That's the real reason that this line of code exists.